Unit test connecting UVC host to UVC device

josuah opened the pull request:
This is a draft implementation not meant to be merged, but as a work bench and support for discussion around #93517.

It uses [UVB](https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/usb/uvb) and the virtual USB device and host back-ends to experiment with the USB APIs.

This permit to run the device class and host class without requiring USB hardware.

To build and run on native sim:
```
west build -b native_sim -S video-sw-generator -t run samples/subsys/usb/uvc_uvb
```
<details><summary>output</summary>
<pre><code>[00:00:00.000,000] <inf> udc_vrt: Device 0x8060e98 (max. speed 1) belongs to uhc-virtual
*** Booting Zephyr OS build v4.2.0-559-g5eb38f5daeb4 ***
[00:00:00.000,000] <inf> usbd_init: interface 0 alternate 0
[00:00:00.000,000] <inf> usbd_init: interface 1 alternate 0
[00:00:00.000,000] <inf> usbd_init:     ep 0x81 mps 0x0040 interface ep-bm 0x00020000
[00:00:00.000,000] <inf> usbd_init: Instance iface-bm 0x00000003 ep-bm 0x00020000
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #1 for 'RGB3'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #2 for 'YUYV'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #3 for 'RGBP'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #4 for 'BX24'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #5 for 'RGGB'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #6 for 'GRBG'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #7 for 'BA81'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_uvc: Adding format descriptor #8 for 'GBRG'
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #1 for 64x64
[00:00:00.000,000] <inf> usbd_uvc: Adding frame descriptor #2 for 1920x1080
[00:00:00.000,000] <inf> usbd_init: Init class node 0x8065264, descriptor length 943
[00:00:00.000,000] <inf> usbd_init: bNumInterfaces 2 wTotalLength 952
[00:00:00.000,000] <inf> usbd_init: FS bNumConfigurations 1
[00:00:00.000,000] <inf> udc: ep 0x80 is not halted|disabled
[00:00:00.000,000] <inf> usbd_core: Actual device speed 1
[00:00:00.070,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.070,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x00, len 0, s:0 d:0 s:1
[00:00:00.070,000] <inf> usbd_ch9: s-in-status finished
[00:00:00.080,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.080,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 0, s:0 d:0 s:1
[00:00:00.080,000] <inf> usbd_ch9: s-in-status finished
[00:00:00.090,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.090,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x80, len 0, s:0 d:0 s:1
[00:00:00.090,000] <inf> usbd_ch9: s-(out)-status finished
[00:00:00.090,000] <inf> usbh_dev: New device with address 1 state 2
[00:00:00.100,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.100,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 0, s:0 d:0 s:1
[00:00:00.100,000] <inf> usbd_ch9: s-in-status finished
[00:00:00.110,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.110,000] <inf> usbd_ch9: Set Configuration Request value 1
[00:00:00.110,000] <inf> usbd_iface: No endpoints, skip interface
[00:00:00.110,000] <inf> usbd_iface: Modify interface 1 ep 0x81 by op 1 ep_bm 20000
[00:00:00.110,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x80, len 0, s:0 d:0 s:1
[00:00:00.110,000] <inf> usbd_ch9: s-(out)-status finished
[00:00:00.120,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.120,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 0, s:0 d:0 s:1
[00:00:00.120,000] <inf> usbd_ch9: s-in-status finished
[00:00:00.120,000] <inf> usbh_dev: Configuration 1 bNumInterfaces 2
[00:00:00.130,000] <inf> usbd_ch9: Handle control 0x806b440 ep 0x00, len 8, s:1 d:0 s:0
[00:00:00.130,000] <inf> usbd_ch9: Handle control 0x806b464 ep 0x00, len 0, s:0 d:0 s:1
[00:00:00.130,000] <inf> usbd_ch9: s-in-status finished
[00:00:00.130,000] <inf> app_main: buf
                                   12 01 00 02 ef 02 01 40  e3 2f 99 99 02 04 01 02 |.......@ ./......
                                   00 01                                            |..
</code></pre></details>

Commit:
tests: usb: uvc: connect USB device and host via UVB

Add a test to run the USB Video Class host support by using
the existing Zephyr USB Video Class device support.
This allows running implementing the host side from the device side.
A draft implementation of UVC is added leveraging this test.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
The way this works:

1. USB UVC Device and Host are added to the sample and configured
2. USB UVC Device and Host are enabled
3. USB Host currently goes through the enumeration steps in this function:
    https://github.com/zephyrproject-rtos/zephyr/blob/2a88cb50f57cf010fc7d60b7e686f960417d6fd0/subsys/usb/host/usbh_core.c#L71
4. This function was not doing anything at the end, so a new placeholder function temporarily called `usbh_class_init()` is added:
  https://github.com/zephyrproject-rtos/zephyr/blob/5eb38f5daeb4c9f7c7192ef4a21654a89dd851e6/subsys/usb/host/usbh_core.c#L77
5. The implementation is for now a placeholder that directly parses the UVC class regardless, without implementing a real USB host class support:
   https://github.com/zephyrproject-rtos/zephyr/blob/5eb38f5daeb4c9f7c7192ef4a21654a89dd851e6/samples/subsys/usb/uvc_uvb/src/main.c#L23

The UVC class implementation can now start, before the USB Host class API is present.

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
Thank you for your answer @MarkWangChinese.

> get audio stream from one commercial usb camera

If I understand correctly, you would be working on an USB Audio Class (such as UAC2) support?

It seems like we have a skeleton of functions in `usbh.h` to implement a class, which I reused in a stub [`usbh_uvc.c`](https://github.com/zephyrproject-rtos/zephyr/pull/93666/files#diff-238c4e04f6752a970a2b2bc98977a0130e5e34062684962f364e3df325e07aa8).

https://github.com/zephyrproject-rtos/zephyr/blob/5c7c3980eb0600a058d4355e3d101078f88d26dc/include/zephyr/usb/usbh.h#L75-L97

This current PR is turned into an unit test.
For now, it is empty, only the twister configuration and basic code layout:

```
west twister -p native_sim -T tests/subsys/usb/uvc
```

Commit:
usb: host: ensure device configuration descriptors are nil-terminated

Allocate extra room at the end of the USB descriptor buffer to ensure
that the device ends with `desc->bLength == 0`. For now, validating
`desc->bLength` from the device to ensure no access past the buffer
is still not done.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: uvc: loop through every descriptor

Loop through each of the VideoStreaming and VideoControl descriptor
to parse them.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

MarkWangChinese commented:
> If I understand correctly, you would be working on an USB Audio Class (such as UAC2) support?
Yes, we are working on the usb host video class and usb host stack with NXP platforms, and connect with actual usb camera.
Our plan is: Firstly writing the basic codes and create the draft pr; secondly plugging usb camera, enumerating and transferring the video stream.

Commit:
usb: host: add host status to the host context

Add a "struct usbh_status" that contains a bitmask of flags to keep
track of the global state of the host context, like done for the
device_next implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: class: implement more of the class/data struct fields

Add a "struct usbh_class_api" for the host implementation, and move all
the function poitners to it. Add more fields to "struct usbh_class_data".

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce wrappers to access the class function pointers

Add API wrappers around the function pointers in struct usbh_class_api,
while also documenting the USB host class internal API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce usbh_class with init/remove functions

Add functions to probe/remove all classes as part of a new usbh_class.c
and a matching usbh_class.h. These functions are called from the function
usbh_init_device_intl() in usbh_core.c to initialize every class upon
connection of a device. The class instances are registered and connected
together as a linked list.

Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: implement class filtering and descriptor browsing

Add utilities to filter a class and search the next descriptor of
a given type. This can be used to in combination to seek device
information from the USB descriptors and try to match a host class
instance given a set of filters.

Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
uvc: device: move helper functions to a common directory

Move UVC helper functions to a file shared between UVC host and device.
The arrays are not visible anymore from either USB host or device, but
instead accessed through a front-end funciton.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: uvc: prepare the header to be used for both host and device

When USB device is not defined, there is no macro for the maximum number
of dwFrameInterval[] defined, so default to 1 instead of
CONFIG_USBD_VIDEO_MAX_FRMIVAL.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
tests: usb: host: test basic API functionality

Add tests making sure the USB Host class APIs introduced build
and run as expected.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
This is superseded by https://github.com/zephyrproject-rtos/zephyr/pull/94591
